File: IntersectionStrategy.py
from typing import List
import numpy as np

from ready_trader_go import Instrument, Lifespan
from MarketState import MarketState
from TradingStrategy import TradingStrategy

import logging
LOT_SIZE_IS = 30
LIFESPAN_IS: Lifespan = Lifespan.FILL_AND_KILL
FEE = 0.0002

class IntersectionStrategy(TradingStrategy):

    def __init__(self):
        super().__init__()
        self.logger = logging.getLogger("DIFFERENCES")
        self.differences_positive = []
        self.differences_negative = []
        self.possible_threshold = np.arange(100, 700, 100)
        self.avg_gain_up = np.ones(10) * 250
        self.tau_up = 400
        self.avg_gain_down = np.ones(10) * 250
        self.tau_down = -400
        self.localMax = 0
        self.localMin = 0
        self.haveSell = False
    def calc_up_threshold(self,last_max):
        '''
        Called everytime a new local max has been recived.
        '''
        for i, t in enumerate(self.possible_threshold):
            if self.tau_up > last_max:
                self.avg_gain_up[i] = self.avg_gain_up[i] * 9 / 10
            else:
                self.avg_gain_up[i] = (self.avg_gain_up[i] * 9 + (t - FEE * 1400))/10
        self.tau_up = (np.argmax(self.avg_gain_up)+1) * 100 #il valore da usare per canSell è questo

    def calc_down_threshold(self,last_min):
        '''
        Called everytime a new local min has been recived.
        '''
        for i, t in enumerate(self.possible_threshold):
            if self.tau_down < last_min:
                self.avg_gain_down[i] = self.avg_gain_down[i] * 9 / 10
            else:
                self.avg_gain_down[i] = (self.avg_gain_down[i] * 9 + (t - FEE * 1400)) / 10
        self.tau_down = -(np.argmax(self.avg_gain_down) + 1) * 100  ##il valore da usare per canBuy è questo

    def canSell(self, instrument: int, etf: MarketState, fut: MarketState) -> bool:  #canBuy e canSell vanno chiamate assieme
        delta =etf.getMean() - fut.getMean()
        self.localMax = max(self.localMax,delta)
        if delta == 0 and self.localMax > 0:
            self.calc_up_threshold(self.localMax)
            self.localMax = 0
        self.logger.warning(self.tau_up)
        return delta >= self.tau_up

    def canBuy(self, instrument: int, etf: MarketState, fut: MarketState) -> bool:
        delta = etf.getMean() - fut.getMean()
        self.localMin = min(self.localMin,delta)
        if delta == 0 and self.localMin < 0:

            self.calc_down_threshold(self.localMin)
            self.localMin = 0
        self.logger.warning(self.tau_down)
        return delta <= self.tau_down

    def calcAskSettings(self, instrument: int, etf: MarketState, fut: MarketState):
        if instrument == Instrument.ETF:
            return [etf.getMaxBid(), LOT_SIZE_IS, LIFESPAN_IS]
        else:
            return [fut.getMaxBid(), LOT_SIZE_IS, LIFESPAN_IS]

    def calcBidSettings(self, instrument: int, etf: MarketState, fut: MarketState):
        if instrument == Instrument.ETF:
            return [etf.getMinAsk(), LOT_SIZE_IS, LIFESPAN_IS]
        else:
            return [fut.getMinAsk(), LOT_SIZE_IS, LIFESPAN_IS]

File: cpp/cppready_trader_go/IntersectionStrategy.cpp
//
// Created by ale on 16/03/23.
//
#include <bits/stdc++.h>
#include "libs/ready_trader_go/types.h"
#include "MarketState.cpp"

using namespace std;



const unsigned int LOT_SIZE_IS = 30;
const Lifespan LIFESPAN_IS = Lifespan::FILL_AND_KILL;
const int THRESHOLD_IS = 400;


class IntersectionStrategy {

private:
    vector<double> avg_gain_up;
    int tau_up = 400;
    vector<double> avg_gain_down;
    int tau_down = -400;
    int lastMax;
    int lastMin;
    mutex updateMutex;

public:

    IntersectionStrategy(){
        avg_gain_up = vector<double>(7,350);
        avg_gain_down = vector<double>(7, 350);
    }

    int argo_max(vector<double> const& vec){
        return distance(vec.begin(), max_element(vec.begin(), vec.end()));
    }

    void calc_up_threshold(int last_diff_up,int meanETF) {
        updateMutex.lock();
        for (int i=0; i<7; i++) {
            int t = (i+1)*100;
            if (t >= last_diff_up) {
                avg_gain_up[i] = avg_gain_up[i] * 5.0 / 6.0;
            } else {
                avg_gain_up[i] = (avg_gain_up[i] * 5.0 + (t - 0.0002 * (meanETF/100))) / 6.0;
            }
        }
        tau_up = (argo_max(avg_gain_up)+1) * 100;
        updateMutex.unlock();
    };

    void calc_down_threshold(int last_diff_down, int meanETF) {
        updateMutex.lock();
        for (int i=0; i<7; i++) {
            int t = -(i+1)*100;
            if (t <= last_diff_down) {
                avg_gain_down[i] = avg_gain_down[i] * 5.0 / 6.0;
            } else {
                avg_gain_down[i] = (avg_gain_down[i] * 5.0 + (abs(t) - 0.0002 * (meanETF/100))) / 6.0;
            }
        }
        tau_down = -(argo_max(avg_gain_down)+1) * 100;
        updateMutex.unlock();
    };

    bool canSell(Instrument instrument,
                 MarketState &etf,
                 MarketState &fut){
        signed int delta = etf.getMean() - fut.getMean();
        updateMutex.lock();
        lastMax = max(lastMax,delta);
        if(delta == 0 and lastMax > 0){
            updateMutex.unlock();
            calc_up_threshold(lastMax,etf.getMean());
            updateMutex.lock();
            lastMax = 0;
        }
        updateMutex.unlock();
        cout<<"tau_up: "<<tau_up<<endl;
        return delta >= max(tau_up,400);
    }
    bool canBuy(Instrument instrument,
                 MarketState &etf,
                 MarketState &fut){
        signed int delta = etf.getMean() - fut.getMean();
        updateMutex.lock();
        lastMin = min(lastMin,delta);
        if(delta == 0 and lastMin < 0){
            updateMutex.unlock();
            calc_down_threshold(lastMin,etf.getMean());
            updateMutex.lock();
            lastMin = 0;
        }
        updateMutex.unlock();
        cout<<"tau_down: "<<tau_down<<endl;
        return delta <= min(tau_down,-400);
    }
    void calcAskSettings(Instrument instrument,
                         MarketState &etf,
                         MarketState &fut,
                         unsigned int &newPrice,
                         unsigned int &volume,
                         Lifespan &lifeSpan){
    if(instrument == Instrument::ETF){
        newPrice = etf.getMaxBid();
    }else{
        newPrice = fut.getMaxBid();
    }
    volume = LOT_SIZE_IS;
    lifeSpan = LIFESPAN_IS;
    }

    void calcBidSettings(Instrument instrument,
                         MarketState &etf,
                         MarketState &fut,
                         unsigned int &newPrice,
                         unsigned int &volume,
                         Lifespan &lifeSpan){
        if(instrument == Instrument::ETF){
            newPrice = etf.getMinAsk();
        }else{
            newPrice = fut.getMinAsk();
        }

        volume = LOT_SIZE_IS;
        lifeSpan = LIFESPAN_IS;
    }

};

